VultureProject/darwin
# VAST/VAML:: ADD SKELETON FOR CMAKE AND TESTS, ADAPT FBUFFER

# :sparkles: Pull Request Template
:bangbang: Once all the **checklist** is **done** you have to:
  * **stash merge** this pull request
  * **delete** the corresponding **branch**
  * **close** the associated **issue**

## :page_with_curl: Type of change

**New feature**: non-breaking change which adds functionality.

## :bulb: Related Issue(s)

- None

## :black_nib: Description
### CMakeLists
- add VAST file inclusion
- add VAML file inclusion
- add option to activate coverage flags

### Functional tests
- add mock test file
- change tests to include new 'float' input type in entries
- update anomaly test to adapt to new algorithm
- add tests for new 'sum' filter
  - test with one value sent
  - test with several (negative, positive, integer) values to test correct summing
  - test with sum under 'required_log_lines'
  - test with existing correct type key overrode at startup
  - test with existing incorrect type key, preventing filter to start

### Test framework
- [FEATURE] add capability to check lines are present in logs
- [FEATURE] add an ApproxDict class to check float equality with delta in dictionaries
- [FEATURE] add a generic filter function to validate fields in generated filters' alerts
- Reuse MOCK_PATH into TEST_FILES_DIR to specify runtime temp dir
- improve Valgrind print returns
- remove useless sleep after stop
- add mock test file for VAML/VAST

### TEST filter
- allow ftest to get lists AND simple values
- simple values are interpreted as string, but can also be numbers

### BUFR
- add 'float' type to valid input formats
- [BREAKING] buffer checks keys in Redis, won't start if key is of wrong type and will delete the key to avoid using stale/improper data
- [NEW] add 'sum' buffer
  - increments a redis key to calculate sum over a time span
  - can take positive and negative values, represented as integers or floats (handled as floats internally)
  - 'required_log_lines' represent a limit under which not to use the absolute rounded value of the sum
  - value is never reinserted in case of error -> won't stack sums in different time intervals

### Dockerfile
- build boost:date_time (required for proprietary VAML)

## :dart: Test Environments

### HardenedBSD (12.2)
- XGBoost (1.2.1)
- Armadillo (10.1.1)
- Redis (6.0.10)
- Boost (1.72.0)
- clang++ (or g++) (10.2.0)
- CMake (3.19.3)
- Python (3.7.9)

### Ubuntu (18.04 docker)
- XGBoost (1.3.1)
- Armadillo (9.900.x)
- Redis (4.0.9)
- Boost (1.71.0)
- g++ (or clang++) (7.5.0)
- CMake (3.17.0)
- Python (3.6.9)
- Valgrind (3.13.0)

## :heavy_check_mark: Checklist:

- [x] (**If other changes**) I have made corresponding changes to the documentation

</br>

- [x] :raising_hand: **I certify on my honor that all the information provided is true, and I've done all I can to deliver a high quality code**
